### DISCUSSION TITLE
Dev build: CSS sometimes (rarely) includes CSS that should not be there

### DISCUSSION THREAD

**Discussion by crummy** [vikejs/vike#954] [Discussion]
Sometimes, I think after a while of working in dev mode, we (me and one other team member) have noticed that CSS from other places in our project ends up imported into a page. Restarting vite seems to resolve the issue.

Roughly this is our setup and what I see:

* `src/*/index.tsx` has entrypoints that we generate with a regular `vite build`
* `ssg/` is our Vike folder
* We run `vite --config vite.ssg.config.ts` to dev
* We build our entrypoint JS files with `vite build`
* We build our Vike SSG pages with `vite build --config vite.ssg.config.ts`

Our config files look like this:
```
// vite.base.config.ts
import react from "@vitejs/plugin-react-swc";
import checker from "vite-plugin-checker";
import * as path from "path";
import type { UserConfigExport } from "vite";

// https://vitejs.dev/config/
export default {
  plugins: [
    react(),
    checker({
      typescript: true,
      eslint: { lintCommand: "eslint . --ext .js,.jsx,.ts,.tsx" },
    }),
  ],
  resolve: {
    alias: {
      ssg: __dirname + "/ssg",
      src: __dirname + "/src",
      [...]
    },
  },
  server: {
    port: 5173,
    proxy: {
      "/services": "http://localhost:4568",
      "/coder": "http://localhost:4568",
      [...]
    },
  },
} satisfies UserConfigExport;
```

```
// vite.config.ts
import type { UserConfigExport } from "vite";
import baseConfig from "./vite.base.config";

export default {
  ...baseConfig,
  build: {
    manifest: true,
    outDir: "out",
    sourcemap: true,
    rollupOptions: {
      input: {
        coder: __dirname + "/src/coder/index.tsx",
        [...]
      },
      output: {
        entryFileNames: "assets/[name].js",
        assetFileNames: "assets/[name].[ext]",
      },
    },
  },
} satisfies UserConfigExport;
```
```
// vite.ssg.config.ts
import type { UserConfigExport } from "vite";
import baseConfig from "./vite.base.config";
import ssr from "vite-plugin-ssr/plugin";

export default {
  ...baseConfig,
  plugins: [...baseConfig.plugins, ssr({ prerender: true })],
  build: {
    outDir: "ssg/out",
  },
} satisfies UserConfigExport;
```
Noticing when this issue occurs is tricky, but sometimes it is obvious because it turns the background colour of our homepage black. Here's my reconstruction:

1. Visit localhost:5173/ (which loads `ssg/pages/index/+Page.tsx`)
2. Notice a black background
3. Open Chrome dev tools and notice CSS imported from `src/coder/index.less`

This is a relatively minor issue but seems a little concerning. I don't have any reliable repro but I'm happy to do some diagnostics next time it happens.

**Comment by brillout** [comment]
I believe that's expected? If you load some CSS and you client-side route to another page, then the CSS is still around. If you refresh the page then the CSS of the previous page is gone. Even though CSS is code-split, it's best to treat each line of CSS as applying to all pages.

**Comment by crummy** [comment]
If I visit:
1. http://localhost:5173 (which loads `ssg/index/+Page.tsx`), then click on navigation to visit
2. http://localhost:5173/coder (which is proxied to our Java backend which includes `src/coder/index.tsx`)
3. then click Back

I don't see any polluted CSS.

**Reply by brillout** [reply]
If you use client-routing then any CSS should be persisted upon client-side navigation.

**Reply by crummy** [reply]
The docs say:
```
By default vite-plugin-ssr does Server Routing. We can opt into Client Routing by setting the [clientRouting option](https://vite-plugin-ssr.com/clientRouting) option to true.
```
We don't have `clientRouting` set to true. And if we did somehow, this navigation is from a VPS page to a non-VPS page so I can't imagine it would apply.

**Reply by brillout** [reply]
In that case your original post is quite surprising. If I can get my hands on a minimal reproduction, then I'll have a look at it. (Although, to be honest, I suspect some kind of user land issue.)

**Reply by crummy** [reply]
I reproed it again, but it only appears maybe once a month (wild guess). Here's a HAR of it happening: https://wormhole.app/N4DPk#H_sG7sO3K7LiI7tXJQ-f9g

Again this time I saw it visiting localhost:5173, though this time it was "infected" by freetrial.less which is not imported anywhere on our home page. I stopped and restarted the Vite process and the CSS returned to normal. Network tab showed no freetrial.less.

edit: just to be clear, I don't expect a fix for this, or even any investigation really - it seems extremely hard to track down. Just commenting to document the behaviour in case it helps down the line.

**Reply by brillout** [reply]
I don't think I can be much of an help without a minimal reproduction. Also the fact that you're the only one reporting this makes me think it's probably unrelated to vite-plugin-ssr.

>  Just commenting to document the behaviour in case it helps down the line.

👍 Keep us updated.
